File: S3/python/billing.py
import boto
import csv


def progress(sofar, total):
    print 'Read %d bytes out of %d' % (sofar, total)

    
class AwsBilling(object):
    def __init__(self, config_csv):
        with open(config_csv, 'rb') as config:
            reader = csv.DictReader(config)
            row = reader.next()
            self.s3 = boto.connect_s3(row['key_id'], row['key'])
            self.bucket_name = row['bucket_name']
    
    def get_billing_file(self):
        bucket = self.s3.get_bucket(self.bucket_name)
        csv_zip_name = None
        for key in bucket.list():
            if key.key[-7:] == 'csv.zip':
                csv_zip_name = key.key
                with open(key.key, 'wb') as csv_zip:
                    key.get_file(csv_zip, cb=progress)
                break
        self.s3.close()

        import zipfile

        if zipfile.is_zipfile(csv_zip_name):
            archive = zipfile.ZipFile(csv_zip_name)
            for item in archive.infolist():
                archive.extract(item)
                
        self.csv_name = item.filename
        return self
        
    def analyze(self):
        with open(self.csv_name, 'rb') as csvfile:
            reader = csv.DictReader(csvfile)
            for row in reader:
                if row['RecordType'] == 'StatementTotal':
                    print "%s: $%s\n" % (row['ItemDescription'], row['Cost'])

                
if __name__ == '__main__':
    AwsBilling('config.csv').get_billing_file().analyze()

File: S3/ruby/billing.rb
require "aws-sdk-core"
require "csv"
require "openssl"

class AwsBilling
    def initialize(config_csv)
        # read credentials, region, and other info from the config file
        @config = CSV.table(config_csv)[0]

        creds = Aws::Credentials.new(@config[:key_id], @config[:key])
        @s3 = Aws::S3::Client.new(credentials: creds, region: @config[:region])
    end
    
    def get_billing_file
        csv_zip_name=nil
        bucket_name = @config[:bucket_name]

        resp = @s3.list_objects(bucket: bucket_name)
        resp.contents.each do |object|
            if object.key[-7..-1] == 'csv.zip'
                csv_zip_name = object.key
                break
            end
        end

        @s3.get_object(
            response_target: csv_zip_name,
            bucket: bucket_name,
            key: csv_zip_name
        )

        require "zip"

        @csv_name=csv_zip_name[0..-5]

        zip_file = Zip::ZipFile.open(csv_zip_name)
        zip_file.extract(@csv_name,@csv_name) {true}
        
        return self
    end
    
    def analyse
        billing = CSV.table(@csv_name)
        totals = billing.select {|row| row[:recordtype] == 'StatementTotal'}

        totals.each do |row|
            print "#{row[:itemdescription]}: $#{row[:cost]}\n"
        end
    end
end

# the following overrides SSL verification;
# not good, but otherwise the code does not work
# TODO: need to find a better solution
OpenSSL::SSL::VERIFY_PEER = OpenSSL::SSL::VERIFY_NONE

AwsBilling.new('config.csv').get_billing_file.analyse
